### PR TITLE
Sorted autocomplete

### LINKED ISSUE #1695: Autocomplete in groups doesn't always show exact match
A small but annoying issue, regarding the autocomplete when setting permissions in a repository. SCM-Manager 2.19.1

Situation:
- we have more than 5 groups that begin with the same prefix.
- we have a group whose name is exactly the given prefix.

In the Add New Permission textbox, we write the prefix.
We receive 5 suggestions - but none of them is the group that is named exactly as the prefix.
So it is not possible to find and select the group that we are looking for.

(It is actually possible if we choose the 6th option "Create prefix", but that is completely counter-intuitive, as that group already exists and won't be created)

I attach screenshots to illustrate the problem (the prefix has been changed).

image

image

I would suggest that the suggestion logic checks if there is an exact match, and if so, puts it in the first position. Alternatively, sort the matches first by length and then alphabetically could also work.

I assume that the autocomplete of users might have the same issue - haven't tested it though.

### PR DESCRIPTION
**Pull request by pfeuffer**
## Proposed changes

Users, groups, repositories and repository roles have been sorted in the rest layer by default if no other sort option was given. In the layers "below" (aka the manager classes or the dao), the collections have been unsorted. This led to the effect, that the autocomplete resource, which did not sort all values beforehand, returned unsorted results. As a sideeffect, direct matches for an input could occur at a random position or not at all (as reported in #1695), when there were enough other matches.

With this pull request the databases for users, groups, repositories and repository roles will use instances of `TreeMap` instead of `LinkedHashMap` internally, so that these values are sorted implicitly (by id respectively name for users, groups and repository roles and namespace/name for repositories).

Due to this change the default sort applied in the rest layer could be removed.

Fixes #1695  

### Your checklist for this pull request

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

**Contributor**:
- [x] PR is well described and the description can be used as a commit message on squash
- [x] Related issues linked to PR if existing and labels set
- [x] New code is covered with unit tests
- [x] [Changelog entry file](https://github.com/scm-manager/changelog#changelog-entry-files) created in `gradle/changelog` or CHANGELOG.md is updated for plugins
- [x] Feature has been tested with different permissions

**Reviewer**:
- [ ] The clean code principles are respected ([CleanCode](https://clean-code-developer.com/virtues/))
- [ ] All new code/logic is implemented on the right spot / "Should this be done here?"
- [ ] UI changes fits into the layout
- [ ] The UI views / components are responsive (mobile views)
- [ ] Correct translations are available

### Checklist for branch merge request (not required for forks)

- [ ] Branch is green/blue on [Jenkins](https://oss.cloudogu.com/jenkins/)
- [ ] Quality Gate passed on [SonarQube](https://sonarcloud.io/organizations/scm-manager/projects)

### CODE REVIEW

File: scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java [hunk 1]
```diff
@@ -67,17 +76,20 @@ public XmlRepositoryDAO(PathBasedRepositoryLocationResolver repositoryLocationRe
     this.repositoryExportingCheck = repositoryExportingCheck;
 
     this.byId = new ConcurrentHashMap<>();
```
Comment (sdorra) [inline scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java:78]: Does this map still have to be concurrent?

File: scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java [hunk 2]
```diff
@@ -201,8 +231,44 @@ public Long getLastModified() {
...
+    Lock lock = byNamespaceLock.readLock();
+    lock.lock();
+    try {
+      return runnable.get();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private static class RepositoryContainer {
```
Comment (sdorra) [inline scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java:259]: I don't get the need of the container. Why do we need it? Why don't we just use the repository without this wrapper?

File: scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java [hunk 3]
```diff
@@ -149,11 +170,20 @@ public void modify(Repository repository) {
 
     synchronized (this) {
```
Comment (sdorra) [inline scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java:171]: Do we still have to synchronize? Is the lock not enough?

File: scm-ui/ui-api/src/repositories.ts [hunk 4]
```diff
@@ -153,103 +151,103 @@ export const useDeleteRepository = (options?: UseDeleteRepositoryOptions) => {
         if (options?.onSuccess) {
           options.onSuccess(repository);
         }
-        await queryClient.removeQueries(repoQueryKey(repository));
+        queryClient.removeQueries(repoQueryKey(repository));
```
Comment (sdorra) [inline scm-ui/ui-api/src/repositories.ts:154]: Do we no longer have to wait for the removal?

**Review by sdorra — CHANGES_REQUESTED** [review]

File: scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java [hunk 5]
```diff
@@ -149,11 +170,20 @@ public void modify(Repository repository) {
...
       // remove old namespaceAndName from map, in case of rename
-      Repository prev = byId.put(clone.getId(), clone);
-      if (prev != null) {
-        byNamespaceAndName.remove(prev.getNamespaceAndName());
+      RepositoryContainer container = byId.get(clone.getId());
+      if (container != null) {
+        if (container.getRepository().getNamespaceAndName() != clone.getNamespaceAndName()) {
```
Comment (pfeuffer) [inline scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java:175]: `equals`!

Comment (pfeuffer) [reply scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java]: This emarrasing code block has been removed again

File: scm-ui/ui-api/src/repositories.ts (hunk 4 above)
Comment (pfeuffer) [reply scm-ui/ui-api/src/repositories.ts:154]: No. `queryClient.removeQueries` does not return a `Promise`.

File: scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java (hunk 3 above)
Comment (pfeuffer) [reply scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java]: Rewrote synchronization once more.

File: scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java (hunk 2 above)
Comment (pfeuffer) [reply scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java]: It's gone...

File: scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java (hunk 1 above)
Comment (pfeuffer) [reply scm-dao-xml/src/main/java/sonia/scm/repository/xml/XmlRepositoryDAO.java]: No, rewrote synchronization.

**Pull request merged**